### DISCUSSION TITLE
Cannot get caches to clear by using URIs

### DISCUSSION THREAD

**Discussion by roberttolton** [spatie/laravel-responsecache#372] [Q&A]
Hi, thanks for the work on this package.

I'm using it in a new Laravel project, but am having trouble getting caches to clear with using URIs.

Clearing caches with tags works fine, but I'd like to understand what I am doing wrong that means giving an array of URIs does not clear the caches for those responses.

My routes api.php:

```
Route::domain(config('app.domains.api'))->group(function () {

    Route::get('/stories', [
        \App\Http\Controllers\Api\Story\StoryController::class,
        'index'
    ])->middleware([
        'cacheResponse',
    ]);

});
```

Line used to clear caches that is expected to work:

```
ResponseCache::selectCachedItems()->forUrls('/api/stories')->forget();
```

But this does not clear the cache. I have also tried this which doesn't work either:

```
ResponseCache::forget(['/stories', '/api/stories', 'my-api.lndo.site/api/stories']);
```

Any ideas what I'm doing wrong?

I've edited the `laravel-responsecache` source files with debugging calls to `ray()` and looked at the output, but I cannot see where or what is going wrong. It's like the URIs I'm asking to clear just can't be found in the cache.

Thanks for the help.

**Comment by lgrassini** [comment]
Same here. We want to invalidate cache for a given URI by doing: `ResponseCache::forget('/v1/pages/9/id/x8i3Wv');` but the cache does not get invalidated.

**Comment by lgrassini** [comment]
@roberttolton Have you found a solution for this? Thanks!

**Comment by roberttolton** [comment]
@lgrassini The way I've gotten around it, is to tag my caches like so:

```
Route::get('/stories/setup', [
    \App\Http\Controllers\Api\Story\StoryController::class,
    'setup',
])->middleware([
    'cacheResponse:stories,stories-setup',
]);

Route::get('/stories', [
    \App\Http\Controllers\Api\Story\StoryController::class,
    'index',
])->middleware([
    'cacheResponse:stories,stories-index',
]);
```

And then clear them like so:

```
ResponseCache::clear(['stories', 'stories-setup', 'stories-index']);
```

**Comment by roberttolton** [comment]
Sorry to bother @freekmurze but even to this day I cannot get cache-clearing by URLs to work. Is there a hidden gotcha?

I'm using the default suffix, so either User ID or empty string, and have the following URL:

```
/api/dashboards?order_by=created_at&order_direction=desc&page=1&per_page=100
```

This does not get cleared by any of these calls:

```
ResponseCache::selectCachedItems()->forUrls('/api/dashboards')->forget();
ResponseCache::selectCachedItems()->forUrls('/dashboards')->forget();
ResponseCache::selectCachedItems()->forUrls('dashboards')->forget();
```

**Reply by freekmurze** [reply]
I'd accept a PR that fixes this.

**Comment by jimuelpalaca** [comment]
Hi @roberttolton @freekmurze , I figure out what's happening with the cache-clearing by URLs. The value of the `$request`'s host created on the `build` function of `AbstractRequestBuilder` is always `localhost`, which doesn't match the host of the request coming from the routes. It affects the key hashing of the response cache.

I tried using the `withRemoteAddress(config('app.url'))` but still doesn't affect the host of the `$request` in `AbstractRequestBuilder`.

Workaround: 
- Append the `config('app.url')` on every URLs you pass on `forUrls`, make sure your `APP_URL` is set on `.env`.

```
ResponseCache::selectCachedItems()
    ->usingSuffix($request->user()->id)
    ->forUrls(config('app.url').'/api/profile')
    ->forget();
```

Proposed Solutions:
- Append the `config('app.url')` when building a new Request on `AbstractRequestBuilder`
- Or, add a `setHost` function to pass the value of the host, then append it when creating a new request.

**Reply by roberttolton** [reply]
Thanks @jimuelpalaca ! I'll try and confirm this in a project when I am able to, and submit a PR if not done yet.

**Reply by apeisa** [reply]
@jimuelpalaca @roberttolton did you work on this any further? I am hitting this same problem and looking into fixing this unless you have already done some work on this.

**Reply by roberttolton** [reply]
@apeisa Haven't gotten around to it yet, need to finish another PR first - so you're welcome to have a look at this.

**Reply by apeisa** [reply]
Here is the PR https://github.com/spatie/laravel-responsecache/pull/409

It works for me. I had to tweak tests, but it actually made sense, since the tests tried to cache spatie.be domain.

**Reply by apeisa** [reply]
And thanks to @jimuelpalaca for the proposed solution. Hopefully, this gets merged, I did spend way too much time initially wondering why cache doesn't invalidate on production.

**Reply by apeisa** [reply]
This is now merged. Hopefully it fixes it for you too.

Important to note though that if just uri provided `forget('/foo')` will only empty it for the public (not individual user caches).

**Comment by caocanhlinh** [comment]
I had this error and got a fix. When you login, it will generate your own user id, so when you clear the cache by url, it will only work for you.

My way will not return a separate cache for each user, instead all users will share the same cache.

(In the file BaseCacheProfile.php, in the function useCacheNameSuffix I return an empty string that is resolved.)
